profiling/src/profiling.rs

```diff
@@ -0,0 +1,806 @@
+    if !module.is_empty() {
+        module.iter().for_each(|byte| buffer.push(*byte));
+        buffer.push(b'|');
+    }
+
+    if !class_name.is_empty() {
+        class_name.iter().for_each(|byte| buffer.push(*byte));
+        buffer.push(b':');
+        buffer.push(b':');
+    }
+
+    // eprintln!("{:?}|{:?}::{:?}", module, class_name, method_name);
+
+    method_name.iter().for_each(|byte| buffer.push(*byte));
+
+    buffer_try_into_string(buffer)
+}
+
+unsafe fn buffer_try_into_string(buffer: Vec<u8>) -> Result<String, Utf8Error> {
+    // Use from_utf8 to verify UTF8; unused Ok result.
+    let _ = std::str::from_utf8(buffer.as_slice())?;
+
+    // The buffer is valid UTF8; convert it to a string without reallocating.
+    let mut wrapper = ManuallyDrop::new(buffer);
+
+    // SAFETY: this was verified with std::str::from_utf8 above
+    Ok(String::from_raw_parts(
+        wrapper.as_mut_ptr(),
+        wrapper.len(),
+        wrapper.capacity(),
+    ))
+}
+
+unsafe fn extract_file_name(execute_data: &zend_execute_data) -> Result<String, Utf8Error> {
+    // this is supposed to be verified by the caller
+    if execute_data.func.is_null() {
+        return Ok(String::new());
+    }
+
+    let func = &*execute_data.func;
+    if func.type_ == ZEND_USER_FUNCTION as u8 {
+        let filename = zend_string_to_bytes(func.op_array.filename);
+        if !filename.is_empty() {
+            let mut buffer = Vec::new();
+            // todo: buffer.try_reserve(filename.len())?;
+            buffer.reserve(filename.len());
+            filename.iter().for_each(|byte| buffer.push(*byte));
+            return buffer_try_into_string(buffer);
+        }
+    }
+    Ok(String::new())
+}
+
+unsafe fn extract_line_no(execute_data: &zend_execute_data) -> u32 {
+    // this is supposed to be verified by the caller
+    assert!(!execute_data.func.is_null());
+
+    let func = &*execute_data.func;
+    if func.type_ == ZEND_USER_FUNCTION as u8 && !execute_data.opline.is_null() {
+        let opline = &*execute_data.opline;
+        return opline.lineno;
+    }
+    0
+}
+
+unsafe fn collect_stack_sample(
+    top_execute_data: *mut zend_execute_data,
+) -> Result<Vec<ZendFrame>, Utf8Error> {
+    let mut samples = Vec::new();
+    let max_depth = 128;
+    samples.reserve(max_depth); // todo: try_reserve
+    let mut execute_data = top_execute_data;
+
+    while !execute_data.is_null() && samples.len() < samples.capacity() {
+        let func = (*execute_data).func;
+        if !func.is_null() {
+            let zend_function = &*func;
+            let function_result = extract_function_name(zend_function);
+
+            // ensure that str in Some(str) is non-empty, simplifies code below
+            let function = if let Ok(str) = function_result {
+                if !str.is_empty() {
+                    Some(str)
+                } else {
+                    None
+                }
```

Comment:
This can be simplified, though it is slightly awkward:
```suggestion
                (!str.is_empty()).then(|| str)
```